### PR TITLE
[MLX] Add Metal profiling hooks to server profiler

### PR DESCRIPTION
**Pull request by LijuanTang94**
## Motivation

Extends the Apple Silicon profiling support first added in #22159 (which added MLX Metal capture to `bench_one_batch.py`) to the server-side profiler, enabling Metal profiling for `bench_serving.py` and `bench_one_batch_server.py`.

Related to the [Apple Device Support roadmap](https://github.com/sgl-project/sglang/issues/19137) — Profiling section.

## Modifications

**`hardware_backend/mlx/profiler.py`** (new file)

Implements the Metal profiling backend behind `apply_metal_profiler_patches()`, mirroring the existing NPU patch pattern:

- Patches `torch.profiler.profile` at module load time so `profiler_manager.py` needs no Apple-specific logic
- **MLX mode** (`SGLANG_USE_MLX=1`): uses `mx.metal.start_capture` / `stop_capture`
- **MPS mode** (torch-native on Apple Silicon): uses `torch.mps.profiler.metal_capture`
- Both paths return `ProfileReqOutput(success=False)` with an actionable message on failure (e.g. missing `MTL_CAPTURE_ENABLED=1`) instead of propagating a `RuntimeError`

**`profiler_manager.py`**

- Adds `_is_mps = is_mps()` and a 3-line `elif _is_mps` branch to call `apply_metal_profiler_patches()` alongside the existing `_is_npu` branch
- Wraps `self.torch_profiler.start()` in `try/except RuntimeError` so capture failures return `success=False` gracefully
- Moves `ProfileManager` import to the top of the file

**`test/registered/unit/hardware_backend/mlx/test_metal_profiler.py`** (new file)

10 unit tests covering both paths:
- `apply_metal_profiler_patches()` replaces `torch.profiler.profile` and is idempotent
- MLX path: `start_mlx` success and `RuntimeError` → `success=False`
- MPS path: `start_mps` success and `RuntimeError` → `success=False`
- `SchedulerProfilerManager._start_profile` graceful failure and full start/stop cycle

## Usage

```bash
# Terminal 1 — launch server with Metal capture enabled
MTL_CAPTURE_ENABLED=1 SGLANG_TORCH_PROFILER_DIR=./traces \
  python -m sglang.launch_server \
    --model-path Qwen/Qwen2.5-0.5B-Instruct \
    --port 30000

# Terminal 2 — benchmark with profiling (MLX mode)
SGLANG_USE_MLX=1 python -m sglang.bench_serving \
    --backend sglang --port 30000 \
    --num-prompts 1 --profile

# MPS mode (torch-native, no SGLANG_USE_MLX needed)
python -m sglang.bench_serving \
    --backend sglang --port 30000 \
    --num-prompts 1 --profile
```

## Verification

Verified on **M4 Pro / macOS 26.3.1 / Python 3.11 / mlx 0.31.2**:

**MLX path** (`SGLANG_USE_MLX=1`):
- `torch.profiler.profile` replaced by `MetalTorchProfiler` ✅
- `start()` / `stop()` produce a `.gputrace` file ✅
- Missing `MTL_CAPTURE_ENABLED=1` → `success=False` with actionable message, no crash ✅

**MPS path** (torch-native):
- `torch.mps.profiler.metal_capture` API available ✅
- `start_mps` success and failure paths both handled ✅

**Unit tests**: 10/10 pass ✅

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #27660670447](https://github.com/sgl-project/sglang/actions/runs/27660670447)
Latest PR Test (Extra): :x: [Run #27660670409](https://github.com/sgl-project/sglang/actions/runs/27660670409)

### CODE REVIEW

**Comment by LijuanTang94** [comment]
Hi @yeahdongcn, lint is green — would appreciate a review when you have a moment 🙏

File: python/sglang/srt/managers/scheduler_components/profiler_manager.py [hunk 1]
```diff
@@ -242,6 +243,17 @@ def _start_profile(
                 torch.cuda.cudart().cudaProfilerStart()
             self.profile_in_progress = True
 
+        if use_mlx():
```
Comment (yeahdongcn) [inline python/sglang/srt/managers/scheduler_components/profiler_manager.py]: We already have `is_mps`. Could you investigate further to see whether this can be enabled for both `MPS` and `MLX`?

Comment (LijuanTang94) [reply python/sglang/srt/managers/scheduler_components/profiler_manager.py]: Good call. Updated the condition to `is_mps() and is_mlx_available()` so Metal capture is enabled for both MPS and MLX modes — as long as the `mlx` package is installed, since `mx.metal` is the only capture API available on Apple Silicon. `use_mlx()` was too narrow and excluded users running torch-native MPS without `SGLANG_USE_MLX=1`. PTAL.

Comment (yeahdongcn) [reply python/sglang/srt/managers/scheduler_components/profiler_manager.py]: This does not look clean to me yet. Apple Silicon currently has two separate execution paths: `torch.mps` and `mlx`/`mlx-lm`. The current implementation mixes those paths in the common profiler flow, including capture API selection and generated trace locations, which makes the behavior harder to reason about.

Can we keep the common scheduler profiler code mostly unchanged and move the Apple-specific profiling logic behind a backend patch/hook, similar to the NPU profiler patch? The gate also feels a bit awkward. `is_mps()` is already a platform-level check, but the actual profiling behavior still needs to distinguish `torch.mps` from MLX.

A quick commit for reference using cc: https://github.com/yeahdongcn/sglang/tree/xd/metal_profiler Please take a look.

Comment (LijuanTang94) [reply python/sglang/srt/managers/scheduler_components/profiler_manager.py]: Thanks for the detailed feedback and the reference branch — really helpful to see the backend patch approach in action. I'll take a look at your implementation and rework accordingly.

File: python/sglang/srt/managers/scheduler_components/profiler_manager.py [hunk 2]
```diff
@@ -36,11 +37,15 @@
...
+elif _is_mps:
+    from sglang.srt.hardware_backend.mlx.profiler import apply_metal_profiler_patches
+
+    apply_metal_profiler_patches()
 
 logger = logging.getLogger(__name__)
 
 
-from sglang.srt.utils.profile_utils import ProfileManager
+from sglang.srt.utils.profile_utils import ProfileManager  # noqa: E402
```
Comment (yeahdongcn) [inline python/sglang/srt/managers/scheduler_components/profiler_manager.py]: Should it be moved to top?

File: python/sglang/srt/managers/scheduler_components/profiler_manager.py (hunk 1 above)
Comment (LijuanTang94) [reply python/sglang/srt/managers/scheduler_components/profiler_manager.py]: Done — reworked following your reference. Moved all Apple Silicon profiling logic into `hardware_backend/mlx/profiler.py` behind `apply_metal_profiler_patches()`, mirroring the NPU patch pattern. `profiler_manager.py` now only adds a 3-line `elif _is_mps` branch; the rest of the scheduler code is untouched.

MLX mode uses `mx.metal.start_capture`, MPS mode uses `torch.mps.profiler.metal_capture`. Verified both paths locally. PTAL.

File: python/sglang/srt/managers/scheduler_components/profiler_manager.py (hunk 2 above)
Comment (LijuanTang94) [reply python/sglang/srt/managers/scheduler_components/profiler_manager.py]: Good catch, moved to the top with the other imports.

**Comment by yeahdongcn** [comment]
@jonahbernard Could you please take a look if you have time? Thanks!

File: python/sglang/srt/hardware_backend/mlx/profiler.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        start_metal_capture: Callable[[Path], tuple[Any, ProfileReqOutput]],
+        torch_profiler: Optional[Any] = None,
+    ):
+        self.start_metal_capture = start_metal_capture
+        self.torch_profiler = torch_profiler
+        self.metal_profiler = None
+
+    def start(self):
+        trace_path = _new_temp_gputrace_path()
+        self.metal_profiler, result = self.start_metal_capture(trace_path)
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/profiler.py:118]: Both `start_mps` and `start_mlx` can throw `RuntimeError`, so failures during capture start will escape.

File: python/sglang/srt/hardware_backend/mlx/profiler.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+    def stop(self):
+        try:
+            if self.torch_profiler is not None:
+                self.torch_profiler.stop()
+        finally:
+            if self.metal_profiler is not None:
+                self.metal_profiler.stop()
+
+    def export_chrome_trace(self, path: str):
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/profiler.py:136]: Could you test both `mps`/`mlx` paths and update the PR description (may also help others to follow)? Thanks!

File: python/sglang/srt/hardware_backend/mlx/profiler.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+import torch
+
+from sglang.srt.managers.io_struct import ProfileReqOutput
+from sglang.srt.utils.tensor_bridge import use_mlx
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class MetalCaptureProfiler:
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/profiler.py:22]: It would be better if we could add some unit tests for this file.

File: python/sglang/srt/hardware_backend/mlx/profiler.py (hunk 3 above)
Comment (LijuanTang94) [reply python/sglang/srt/hardware_backend/mlx/profiler.py:118]: Fixed — wrapped `self.torch_profiler.start()` in `_start_profile` with `try/except RuntimeError` so failures return `ProfileReqOutput(success=False)` instead of propagating.

File: python/sglang/srt/hardware_backend/mlx/profiler.py (hunk 4 above)
Comment (LijuanTang94) [reply python/sglang/srt/hardware_backend/mlx/profiler.py:136]: Done — verified both MLX and MPS paths locally and updated the PR description with usage instructions and verification results for each path.

File: python/sglang/srt/hardware_backend/mlx/profiler.py (hunk 5 above)
Comment (LijuanTang94) [reply python/sglang/srt/hardware_backend/mlx/profiler.py:22]: Added `test/registered/unit/hardware_backend/mlx/test_metal_profiler.py` with 10 unit tests covering both MLX and MPS paths, patch idempotency, RuntimeError handling, and the full `SchedulerProfilerManager` start/stop cycle. All pass locally.

**Review by yeahdongcn — APPROVED** [review]

**Comment by LijuanTang94** [comment]
Hi @yeahdongcn — `lint` is green now (the failure was an isort import-ordering nit in the test file, fixed in e475d004a). The remaining red `pr-gate`/`*-finish` checks are just the per-user 120-min CI cooldown kicking in after a quick re-push, not real test failures. Could you kick off `/tag-and-rerun-ci` when you have a moment? Thanks! 🙏
